### DISCUSSION TITLE
How to post modal form data to controller?

### DISCUSSION THREAD

**Discussion by OvalMedia** [filamentphp/filament#5469] [Help] [answered]
I have created an action that should trigger a modal with a form for sending an email.

```
Actions\Action::make('email')
	->label('E-Mail')
	->color('success')
	->icon('entypo-mail')
	->requiresConfirmation()
	->modalHeading('Send via mail')
	->modalButton('Send')		
	->form([
		TextInput::make('to')->label('To')->required(),
		TextInput::make('subject')->label('Subject')->required(),
		RichEditor::make('body')->label('')			
	])
	->action(
		function (array $data) {
			$data['invoice'] = $this->data['id'];
			return redirect()->route('invoice.mail', $data);
		}
	)
```

When I submit the modal the data should be sent to a post route (`'invoice.mail'`). But redirect() always creates a get route. How would I create a proper post route in this context?

**Comment by Wulfheart** [comment]
I solve it like this: 
```
->action(fn(Model $record, AddNMRForPlayerAction $actor) => $actor->execute($record))
```

**Comment by danharrin** [comment, marked as answer]
You don't need to create a post route, you can execute the logic within action().

**Reply by OvalMedia** [reply]
There will be a lot of code that I would like to reuse elsewhere. If I put it all in the action that will not be possible.

**Reply by danharrin** [reply]
Create an invokable Action class in app/Actions, and then inject that inside your action() callback. Redirecting to a controller is not performant

**Reply by OvalMedia** [reply]
Thanks. Always learning new stuff. For those who are scratching their heads about this like me, here is a nice introduction to what action classes are and how and why you should use them: https://medium.com/@remi_collin/keeping-your-laravel-applications-dry-with-single-action-classes-6a950ec54d1d

**Reply by Hashim-svg** [reply]
Its not working / and not showing any error . sending data from custom action modal form 


Actions\Action::make('PanelInvoice')->form([
                TextInput::make('companyname')
                    ->label('Company Name'),
                DatePicker::make('from')->default(null),
                DatePicker::make('to')->default(null),
            ])
            ->action(
                function (array $data)
                {
                    $companyName=$data['companyname'];
                    $fromDate=$data['from'];
                    $toDate=$data['to'];
                 return   route('invoice-genrate',[$companyName,$fromDate,$toDate]);
                }
            )

Route::get('/invoice-genrate/{companyName}/{fromDate}/{toDate}', [BillPanelReports::class, 'createInvoice'])->name('invoice-genrate');

public function createInvoice(Request $request)
    {
        $companyName = $request->input('companyname');
        $fromDate = $request->input('from');
        $toDate = $request->input('to');

        dd($companyName,$fromDate,$toDate);

    }
